### DISCUSSION TITLE
Result fold's in gleam

### DISCUSSION THREAD

**Discussion by tearaj** [gleam-lang/gleam#3966] [Questions & support] [answered]
Hey, I am a newbie in gleam, i was trying to solve a scenario where i have a result object and based on whether the result is Ok or Error, i would perform a addition operation, below is the snippet of code

```
  let occurances = get_occurances(num_list)
  use acc, item <- list.fold(num_list, 0)
  case dict.get(occurances, item) {
    Ok(occured_times) -> {
      acc + { item * occured_times }
    }
    _ -> acc
  }
  ```
  
  i was looking into how to change the case dict.get into a piping operation using the result library, something like 
 ```
  dict.get(occurances, item)
    |> result.map(fn(occ_time) { acc + { item * occ_time } })
```

but i was facing issues trying to figure out how to perform the operation during the error case when the key doesnt exist in the Dict. I tried using map_error, but that wraps the output i want inside the Error() and not what i want. This is what i tried
```
    dict.get(occurances, item)
    |> result.map(fn(occ_time) { acc + { item * occ_time } })
    |> result.map_error(fn(_){acc})
```

My question is if there is any way to write the above like
`result.fold(result_object, on_error_function, on_ok_function)`
the above was inspired from the Arrow.kt library in Kotlin.

Please let me know if there is a way to achieve what i want

**Comment by lpil** [comment]
I would write it with a case expression, but without `use`. I believe this would be the easiest to understand code.

**Reply by tearaj** [reply]
Thanks for responding Ipil! 
I get that it would be easier to understand it without the `use`, (i was trying to figure out all possible ways to make use of `use`).  
And i wanted to understand if there is a way to maybe make use of the result library to perform some function if the result is an error and some other if the result is a success, like if there was some built in way to do that without using case expressions

**Reply by lpil** [reply, marked as answer]
You could use `map` with `unwrap` perhaps.

**Reply by tearaj** [reply]
thank you!
